File: ports/clojure/src/sprite_tool/watch.clj
(ns sprite-tool.watch
  "sprite-tool watch: Poll a sprite's beads tracker task for progress."
  (:require [clojure.string :as str]
            [sprite-tool.sprite :as sprite])
  (:import [java.time LocalTime]
           [java.time.format DateTimeFormatter]))

(defn- usage []
  (println "Usage: sprite-tool watch <sprite-name> [task-id] [poll-interval]")
  (println)
  (println "Arguments:")
  (println "  sprite-name     Name of the sprite to watch")
  (println "  task-id         Beads task ID to track (default: auto-detect first open critical task)")
  (println "  poll-interval   Seconds between polls (default: 30)")
  (println)
  (println "Examples:")
  (println "  sprite-tool watch ember-red-hawk")
  (println "  sprite-tool watch ember-red-hawk CRM-1")
  (println "  sprite-tool watch ember-red-hawk CRM-1 60")
  (System/exit 1))

(defn- time-now []
  (.format (LocalTime/now) (DateTimeFormatter/ofPattern "HH:mm:ss")))

(defn- sx-quiet
  "Run command in sprite, suppress stderr, return stdout."
  [sprite-name cmd]
  (let [result (sprite/run-process
                ["sprite" "exec" "-s" sprite-name "bash" "-c" cmd]
                :inherit-io false :capture true)]
    (or (:stdout result) "")))

(defn run
  "Execute the watch subcommand."
  [args]
  (when (empty? args) (usage))

  (let [sprite-name   (first args)
        task-id-arg   (second args)
        poll-interval (if (> (count args) 2)
                        (try
                          (Integer/parseInt (nth args 2))
                          (catch NumberFormatException _
                            (println "Error: poll-interval must be an integer")
                            (System/exit 1)
                            30))
                        30)

        ;; Auto-detect tracker task if not specified
        task-id (if (not (str/blank? task-id-arg))
                  task-id-arg
                  (do
                    (println "Detecting tracker task...")
                    (let [critical (str/trim
                                   (sx-quiet sprite-name
                                             "cd /home/sprite && bd list --priority critical 2>/dev/null | head -1 | awk '{print $1}'"))
                          task (if (str/blank? critical)
                                 (do
                                   (println "No critical task found. Falling back to first open task...")
                                   (str/trim
                                    (sx-quiet sprite-name
                                              "cd /home/sprite && bd list 2>/dev/null | head -1 | awk '{print $1}'")))
                                 critical)]
                      (when (str/blank? task)
                        (println (str "ERROR: No beads tasks found on sprite '" sprite-name "'"))
                        (println (str "Specify a task ID manually: sprite-tool watch " sprite-name " <task-id>"))
                        (System/exit 1))
                      (println (str "Tracking task: " task))
                      task)))]

    (println (str "Watching sprite '" sprite-name "' task '" task-id "' (every " poll-interval "s)"))
    (println "Press Ctrl+C to stop")
    (println)

    (loop []
      ;; Clear screen: \033[2J\033[H
      (print "\033[2J\033[H")
      (flush)

      (println (str "=== sprite-watch: " sprite-name " / " task-id " === " (time-now) " ==="))
      (println)

      ;; Show task status
      (let [task-output (sx-quiet sprite-name
                                  (str "cd /home/sprite && bd show " task-id " 2>/dev/null"))]
        (if (str/blank? task-output)
          (println "(could not read task)")
          (println task-output)))
      (println)

      ;; Show recent comments
      (println "--- Recent updates ---")
      (let [comments (sx-quiet sprite-name
                               (str "cd /home/sprite && bd comments " task-id " 2>/dev/null | tail -8"))]
        (if (str/blank? comments)
          (println "(no comments)")
          (println comments)))
      (println)

      ;; Check if done
      (let [status-line (sx-quiet sprite-name
                                   (str "cd /home/sprite && bd show " task-id " 2>/dev/null | grep -i status"))
            done? (and (not (str/blank? status-line))
                       (re-find #"(?i)closed|done|completed" status-line))]
        (if done?
          (do
            (println "==========================================")
            (println "PROJECT COMPLETE")
            (println "=========================================="))
          (do
            (Thread/sleep (* poll-interval 1000))
            (recur)))))))

File: ports/elixir/lib/sprite_tool/watch.ex
defmodule SpriteTool.Watch do
  @moduledoc """
  sprite-tool watch: Poll a sprite's beads tracker task for progress.
  """

  alias SpriteTool.Sprite

  @doc """
  Execute the watch subcommand.
  """
  @spec run([String.t()]) :: :ok
  def run(args) do
    case args do
      ["--help" | _] ->
        usage()
        System.halt(0)

      ["-h" | _] ->
        usage()
        System.halt(0)

      [] ->
        usage()
        System.halt(1)

      [sprite_name | rest] ->
        {task_id, poll_interval} = parse_watch_args(rest)
        do_watch(sprite_name, task_id, poll_interval)
    end
  end

  defp parse_watch_args([]) do
    {"", 30}
  end

  defp parse_watch_args([task_id]) do
    {task_id, 30}
  end

  defp parse_watch_args([task_id, interval_str | _]) do
    interval =
      case Integer.parse(interval_str) do
        {n, ""} -> n
        _ -> 30
      end

    {task_id, interval}
  end

  defp do_watch(sprite_name, task_id, poll_interval) do
    # Auto-detect tracker task if not specified
    task_id = resolve_task_id(sprite_name, task_id)

    IO.puts("Watching sprite '#{sprite_name}' task '#{task_id}' (every #{poll_interval}s)")
    IO.puts("Press Ctrl+C to stop")
    IO.puts("")

    watch_loop(sprite_name, task_id, poll_interval)
  end

  defp resolve_task_id(sprite_name, "") do
    IO.puts("Detecting tracker task...")

    # Try critical tasks first
    task_id =
      Sprite.sx(
        sprite_name,
        "cd /home/sprite && bd list --priority critical 2>/dev/null | head -1 | awk '{print $1}'"
      )
      |> normalize_task_id()

    task_id =
      if task_id == "" do
        IO.puts("No critical task found. Falling back to first open task...")

        Sprite.sx(
          sprite_name,
          "cd /home/sprite && bd list 2>/dev/null | head -1 | awk '{print $1}'"
        )
        |> normalize_task_id()
      else
        task_id
      end

    if task_id == "" do
      IO.puts("ERROR: No beads tasks found on sprite '#{sprite_name}'")
      IO.puts("Specify a task ID manually: sprite-tool watch #{sprite_name} <task-id>")
      System.halt(1)
    end

    IO.puts("Tracking task: #{task_id}")
    task_id
  end

  defp resolve_task_id(_sprite_name, task_id), do: task_id

  defp normalize_task_id(nil), do: ""
  defp normalize_task_id(str), do: String.trim(str)

  defp watch_loop(sprite_name, task_id, poll_interval) do
    # Clear screen and move cursor to top-left
    IO.write("\e[2J\e[H")

    now = Calendar.strftime(DateTime.utc_now(), "%H:%M:%S")
    IO.puts("=== sprite-watch: #{sprite_name} / #{task_id} === #{now} ===")
    IO.puts("")

    # Show task status
    task_output =
      Sprite.sx(sprite_name, "cd /home/sprite && bd show #{task_id} 2>/dev/null")

    case task_output do
      nil -> IO.puts("(could not read task)")
      "" -> IO.puts("(could not read task)")
      output -> IO.puts(output)
    end

    IO.puts("")

    # Show recent comments
    IO.puts("--- Recent updates ---")

    comments_output =
      Sprite.sx(sprite_name, "cd /home/sprite && bd comments #{task_id} 2>/dev/null | tail -8")

    case comments_output do
      nil -> IO.puts("(no comments)")
      "" -> IO.puts("(no comments)")
      output -> IO.puts(output)
    end

    IO.puts("")

    # Check if done
    status =
      Sprite.sx(
        sprite_name,
        "cd /home/sprite && bd show #{task_id} 2>/dev/null | grep -i status"
      ) || ""

    if status =~ ~r/closed|done|completed/i do
      IO.puts("==========================================")
      IO.puts("PROJECT COMPLETE")
      IO.puts("==========================================")
      :ok
    else
      Process.sleep(poll_interval * 1000)
      watch_loop(sprite_name, task_id, poll_interval)
    end
  end

  defp usage do
    IO.puts("""
    Usage: sprite-tool watch <sprite-name> [task-id] [poll-interval]

    Arguments:
      sprite-name     Name of the sprite to watch
      task-id         Beads task ID to track (default: auto-detect first open critical task)
      poll-interval   Seconds between polls (default: 30)

    Examples:
      sprite-tool watch ember-red-hawk
      sprite-tool watch ember-red-hawk CRM-1
      sprite-tool watch ember-red-hawk CRM-1 60\
    """)
  end
end
